consensus/src/processes/dagtraversalmanager.rs

```diff
@@ -0,0 +1,147 @@
+    ) -> Self {
+        Self { genesis_hash, ghostdag_store, block_window_cache_store, block_window_cache_store_window_size }
+    }
+    pub fn block_window(
+        &self, high_ghostdag_data: Arc<GhostdagData>, window_size: usize,
+    ) -> BinaryHeap<Reverse<SortableBlock>> {
+        let mut window_heap = SizedUpBlockHeap::new(self.ghostdag_store.clone(), window_size);
+        if window_size == 0 {
+            return window_heap.binary_heap;
+        }
+
+        let mut current_gd = high_ghostdag_data;
+
+        if window_size == self.block_window_cache_store_window_size {
+            if let Some(select_parent_binary_heap) = self
+                .block_window_cache_store
+                .get(&current_gd.selected_parent)
+            {
+                let mut window_heap = SizedUpBlockHeap::from_binary_heap(
+                    self.ghostdag_store.clone(),
+                    window_size,
+                    (*select_parent_binary_heap).clone(),
+                );
+                if current_gd.selected_parent != self.genesis_hash {
+                    self.try_push_mergeset(&mut window_heap, &current_gd);
+                }
+
+                return window_heap.binary_heap;
+            }
+        }
+
+        // Walk down the chain until we finish
+        loop {
+            if current_gd.selected_parent == self.genesis_hash || current_gd.selected_parent.is_origin() {
+                break;
+            }
+
+            let done = self.try_push_mergeset(&mut window_heap, &current_gd);
+            if done {
+                break;
+            }
+
+            current_gd = self
+                .ghostdag_store
+                .get_data(current_gd.selected_parent)
+                .unwrap();
+        }
+
+        window_heap.binary_heap
+    }
+
+    fn try_push_mergeset(&self, heap: &mut SizedUpBlockHeap<T>, ghostdag_data: &GhostdagData) -> bool {
+        let added = heap.try_push(ghostdag_data.selected_parent);
+
+        // If the window is full and the selected parent is less than the minimum then we break
+        // because this means that there cannot be any more blocks in the past with higher blueWork
+        if !added {
+            return true;
+        }
+
+        let mut blues: Vec<Hash> = ghostdag_data.mergeset_blues[1..].to_vec(); // Remove the selected parent
+        blues.reverse(); // Go over the merge set in reverse because it's ordered in reverse by blueWork.
+        for blue in blues.iter().cloned() {
+            let added = heap.try_push(blue);
+
+            // If it's smaller than minimum then we won't be able to add the rest because they're even smaller.
+            if !added {
+                break;
+            }
+        }
+
+        let mut reds = ghostdag_data.mergeset_reds.clone();
+        let reds = Arc::make_mut(&mut reds);
+        reds.reverse(); // Go over the merge set in reverse because it's ordered in reverse by blueWork.
+        for red in reds.iter().cloned() {
```

Comment:
Same here, should be `for red in ghostdag_data.mergeset_reds.iter().rev().cloned()`